src/ppx_deriving_yojson.ml

```diff
@@ -210,147 +210,430 @@ let wrap_runtime decls =
 
 let ser_str_of_type ~options ~path ({ ptype_loc = loc } as type_decl) =
   ignore (parse_options options);
-  let serializer =
-    match type_decl.ptype_kind, type_decl.ptype_manifest with
-    | Ptype_abstract, Some manifest -> ser_expr_of_typ manifest
-    | Ptype_variant constrs, _ ->
-      constrs |>
-      List.map (fun { pcd_name = { txt = name' }; pcd_args; pcd_attributes } ->
-        let args = List.mapi (fun i typ -> app (ser_expr_of_typ typ) [evar (argn i)]) pcd_args in
-        let json_name = attr_name name' pcd_attributes in
-        let result =
-          match args with
-          | []   -> [%expr `List [`String [%e str json_name]]]
-          | args -> [%expr `List ((`String [%e str json_name]) :: [%e list args])]
+  let polymorphize = Ppx_deriving.poly_fun_of_type_decl type_decl in
+  match type_decl.ptype_kind with
+  | Ptype_open ->
+      begin
+        let to_yojson_name = Ppx_deriving.mangle_type_decl (`Suffix "to_yojson") type_decl in
+        let mod_name = Ppx_deriving.mangle_type_decl
+          (`PrefixSuffix ("M_", "to_yojson")) type_decl
         in
-        Exp.case (pconstr name' (List.mapi (fun i _ -> pvar (argn i)) pcd_args)) result) |>
-      Exp.function_
-    | Ptype_record labels, _ ->
-      let fields =
-        labels |>
-        List.mapi (fun i { pld_name = { txt = name }; pld_type; pld_attributes } ->
-          let field  = Exp.field (evar "x") (mknoloc (Lident name)) in
-          let result = [%expr [%e str (attr_key name pld_attributes)],
-                              [%e ser_expr_of_typ pld_type] [%e field]] in
-          match attr_default pld_type.ptyp_attributes with
-          | None ->
-            [%expr [%e result] :: fields]
-          | Some default ->
-            [%expr if [%e field] = [%e default] then fields else [%e result] :: fields])
+        match type_decl.ptype_manifest with
+        | Some ({ ptyp_desc = Ptyp_constr ({ txt = lid }, args) } as manifest) ->
+            let ser = ser_expr_of_typ manifest in
+            let lid = Ppx_deriving.mangle_lid (`PrefixSuffix ("M_", "to_yojson")) lid in
```

Comment:
Because mangle_lid will add one, and I usually generate ids with two underscores to prevent conflicts (I agree that in this case the probability of a conflict is rather low. Same for t__ . Do you prefer only one ?